### PR TITLE
Sparse reset

### PR DESCRIPTION
**Pull request by bpeabody**
Stop using `NodeGit.reset` to reset the meta-repo.

### CODE REVIEW

File: node/lib/util/git_util.js [hunk 1]
```diff
@@ -942,6 +942,7 @@ exports.getUrlFromRemoteName = co.wrap(function *(repo, remoteName) {
 });
 
 /**
+<<<<<<< HEAD
```
Comment (novalis) [inline node/lib/util/git_util.js]: That doesn't look good.

File: node/lib/util/submodule_config_util.js [hunk 2]
```diff
@@ -70,6 +78,106 @@ const CONFIG_FILE_NAME = "config";
...
+    const newConfigLines = [];
+    const searchString = "[submodule \"" + submoduleName + "\"]";
+
+    let found = false;
+    let inSubmoduleConfig = false;
+
+    // Loop through the file and push, onto 'newConfigLines' any lines that
+    // aren't part of the bad submodule section.
+
+    for (let i = 0; i < configLines.length; ++i) {
```
Comment (novalis) [inline node/lib/util/submodule_config_util.js]: for const line of configLines?

File: node/lib/util/submodule_config_util.js [hunk 3]
```diff
@@ -70,6 +78,106 @@ const CONFIG_FILE_NAME = "config";
...
+
+    // If we didn't find the submodule, don't write the data back out.
+
+    if (found) {
+        newConfigLines.push("");  // one more new line
+        fs.writeFileSync(configPath, newConfigLines.join("\n"));
+    }
+
+    // Silence warning about no yield statement.
+    yield (Promise.resolve());
```
Comment (novalis) [inline node/lib/util/submodule_config_util.js]: why not just use the async fs-promise funcs to read the file?

File: node/test/util/close_util.js [hunk 4]
```diff
@@ -39,8 +39,8 @@ const RepoASTTestUtil = require("../../lib/util/repo_ast_test_util");
 describe("close_util", function () {
     describe("close", function () {
         // Don't need to test that deinitialization works, that's tested in
-        // deinit_util; just need to see that we handle paths and dirty
-        // sumobulde situations.
+        // submodule_config_util; just need to see that we handle paths and
+        // dirty sumobulde situations.
```
Comment (novalis) [inline node/test/util/close_util.js]: spelling of submodule

File: node/lib/util/reset.js [hunk 5]
```diff
@@ -69,49 +75,69 @@ function getType(type) {
...
+            }
         }
-        const repo = yield opener.getSubrepo(name);
-        const status = yield StatusUtil.getRepoStatus(repo, {
-            showMetaChanges: true,
+    }
+
+    // Tidy up directories when not in sparse mode.
+
+    if (!(yield SparseCheckoutUtil.inSparseMode(repo))) {
```
Comment (novalis) [inline node/lib/util/reset.js]: If we're in sparse mode, why don't we do this?

File: node/lib/util/git_util.js (hunk 1 above)
Comment (bpeabody) [reply node/lib/util/git_util.js]: how embarrassing!

File: node/lib/util/submodule_config_util.js (hunk 2 above)
Comment (bpeabody) [reply node/lib/util/submodule_config_util.js]: thanks for the reminder

File: node/lib/util/submodule_config_util.js (hunk 3 above)
Comment (bpeabody) [reply node/lib/util/submodule_config_util.js]: I use the synchronous methods when mutating the config file to avoid race conditions.

Comment (bpeabody) [reply node/lib/util/submodule_config_util.js]: (will add a note explaining this)

File: node/lib/util/reset.js (hunk 5 above)
Comment (bpeabody) [reply node/lib/util/reset.js]: Well spell out reason in comment.

**Comment by bpeabody** [comment]
Think I hit everything

**Comment by novalis** [comment]
Are those test-failures expected?

**Comment by bpeabody** [comment]
The test failures are not expected except for the continue one, for which I entered https://github.com/twosigma/git-meta/issues/517 -- I think there is a problem with libgit2 in rendering conflicts but need to do more investigation. I'll dig into the others.

**Comment by bpeabody** [comment]
Nope, that's a different error on the conflict too!

**Comment by bpeabody** [comment]
@novalis -- I cleared up all the tests, one of which did indicate a problem; I was not resetting submodules that were open (potentially with WD/Index changes) but unchanged HEAD. Also, in digging into that I I determined that `ResetUtil.abort` was doing unnecessary work.

**Comment by novalis** [comment]
Conveniently, these test fixes also seem to fix the rm tests.  I'm pushing that rebased on top of this.

**Comment by bpeabody** [comment]
I'm curious what test failures you were seeing with `rm` that this cleared up, @novalis .

**Comment by bpeabody** [comment]
Both are fixed now?

I have seen occasional segfaults too, typically only when I run lots of tests at once. If you can figure out why, that would be awesome.

**Comment by novalis** [comment]
I try to always run the full suite before pushing.  Before these changes, running it would always segfault. After, I got two runs without segfaults (and none with).  And the other one is fixed too.

**Comment by bpeabody** [comment]
Only thing I can guess is that reduced reliance on the `cacheSubmodules` "hack" has helped.

**Pull request merged**